### PR TITLE
Add views for resource group statistics information.

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
* gp_toolkit.gp_resgroup_config provides the resgroup config
  information;
* gp_toolkit.gp_resgroup_status provides the resgroup runtime statistics
  information;
* added new columns to pg_catalog.pg_stat_activity for resgroup per
  activity statistics information;

Some examples:

    -- query resource group config information
    SELECT * FROM gp_toolkit.gp_resgroup_config;

    -- query resource group runtime statistics information
    SELECT * FROM gp_toolkit.gp_resgroup_status;

    -- query statistics information on all activities
    SELECT * FROM pg_catalog.pg_stat_activity;

However as resource group has not been fully implemented yet, so there
is only dummy output at the moment.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@pivotal-ning-yu Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=1852)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=1852) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-ning-yu** [comment]
@xiong-gang @yydzero @eurekaka @guofengrichard @pengzhout Please help to review this PR, thanks in advance.

**Comment by pivotal-issuemaster** [comment]
@pivotal-ning-yu Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=1852)!

File: src/backend/catalog/system_views.sql [hunk 1]
```diff
@@ -376,7 +376,12 @@ CREATE VIEW pg_stat_activity AS
             S.client_port,
             S.application_name,
             S.xact_start,
-            S.waiting_reason
+            S.waiting_reason,
+            S.rsgid,
+            S.rsgname,
+            S.rsgqueueing,
+            S.rsgqueuereason,
+            S.rsgqueueduration
```
Comment (eurekaka) [inline src/backend/catalog/system_views.sql]: Hmm, rsgname, rsgqueueing, rsgqueuereason are not necessary, and rsgqueueduration should be changed to rsgqueuestarttime? user can get rsgname from rsgid, rsgqueueing can be reflected by 'waiting', rsgqueuereason can only be 'max_concurrency'.

File: src/backend/utils/adt/pgstatfuncs.c [hunk 2]
```diff
@@ -424,6 +424,14 @@ pg_stat_get_activity(PG_FUNCTION_ARGS)
...
 			TupleDescInitEntry(tupdesc, (AttrNumber) 13, "waiting_for", TEXTOID, -1, 0);
 
+		if (nattr > 13) {
+			TupleDescInitEntry(tupdesc, (AttrNumber) 14, "rsgid", OIDOID, -1, 0);
+			TupleDescInitEntry(tupdesc, (AttrNumber) 15, "rsgname", TEXTOID, -1, 0);
+			TupleDescInitEntry(tupdesc, (AttrNumber) 16, "rsgqueueing", BOOLOID, -1, 0);
+			TupleDescInitEntry(tupdesc, (AttrNumber) 17, "rsgqueuereason", TEXTOID, -1, 0);
+			TupleDescInitEntry(tupdesc, (AttrNumber) 18, "rsgqueueduration", INTERVALOID, -1, 0);
+		}
+
```
Comment (eurekaka) [inline src/backend/utils/adt/pgstatfuncs.c]: can we remove the if check here?

File: src/backend/utils/adt/pgstatfuncs.c [hunk 3]
```diff
@@ -621,6 +629,19 @@ pg_stat_get_activity(PG_FUNCTION_ARGS)
...
+				MemSet(&interval, 0, sizeof(interval));
+
+				values[13] = ObjectIdGetDatum(0);
+				values[14] = CStringGetTextDatum("default");
+				values[15] = BoolGetDatum(false);
+				values[16] = CStringGetTextDatum("");
+				values[17] = IntervalPGetDatum(&interval);
+			}
+
```
Comment (eurekaka) [inline src/backend/utils/adt/pgstatfuncs.c]: same here.

File: src/backend/utils/adt/pgstatfuncs.c [hunk 4]
```diff
@@ -635,6 +656,14 @@ pg_stat_get_activity(PG_FUNCTION_ARGS)
...
 			if (funcctx->tuple_desc->natts > 12)
 				nulls[12] = true;
+			if (funcctx->tuple_desc->natts > 13)
+			{
+				nulls[13] = true;
+				nulls[14] = true;
+				nulls[15] = true;
+				nulls[16] = true;
+				nulls[17] = true;
+			}
```
Comment (eurekaka) [inline src/backend/utils/adt/pgstatfuncs.c]: and here

File: src/backend/utils/resgroup/resgroup.c [hunk 5]
```diff
@@ -0,0 +1,103 @@
+/*-------------------------------------------------------------------------
+ *
+ * resgroup.c
+ *	  POSTGRES internals code for resource groups.
+ *
```
Comment (yydzero) [inline src/backend/utils/resgroup/resgroup.c]: Shall we use Greenplum instead of POSTGRES?

File: src/backend/utils/resgroup/resgroup.c [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * Copyright (c) 2006-2008, Greenplum inc.
+ * Portions Copyright (c) 1996-2008, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
```
Comment (yydzero) [inline src/backend/utils/resgroup/resgroup.c]: As this is Greenplum added file, so only first line is needed, also need to extend to 2017

File: src/backend/catalog/gp_toolkit.sql [hunk 7]
```diff
@@ -1823,6 +1823,68 @@ AS
...
+    SELECT
+        G.groupid,
+        G.num_running,
+        G.num_queueing,
+        G.cpu_usage,
+        G.memory_usage,
+        G.total_queue_duration,
+        G.total_execution_duration
+    FROM pg_resgroup_get_status(null) as G;
+
```
Comment (eurekaka) [inline src/backend/catalog/gp_toolkit.sql]: cpu_usage is not available I think

**Comment by pivotal-ning-yu** [comment]
Thanks guys for your review. I have updated the code by following most of your comments. The refactoring of function `pg_resgroup_get_status_kv()` is still in progress.

**Comment by pivotal-ning-yu** [comment]
Refactored the function `pg_resgroup_get_status()` with a key-value style one `pg_resgroup_get_status_kv()`.

I'm creating a dummy output for verification.

**Comment by pivotal-ning-yu** [comment]
A dummy output is added for verification. Please help to review again.

In the meanwhile I'll update the testcases to pass the CI check.

**Comment by pivotal-ning-yu** [comment]
BTW I didn't include the changes to catversion.h and 5.json to avoid conflicts.

File: src/backend/catalog/gp_toolkit.sql [hunk 8]
```diff
@@ -1823,6 +1823,82 @@ AS
...
+CREATE VIEW gp_toolkit.gp_resgroup_config AS
+    SELECT
+        G.oid       AS groupid,
+        G.rsgname   AS groupname,
+        T1.value    AS concurrency,
+        T1.proposed AS proposed_concurrency,
+        T2.value    AS cpu_rate_limit,
+        T3.value    AS memory_limit,
+        T3.proposed AS proposed_memory_limit,
+        T4.value    AS redzone
```
Comment (xiong-gang) [inline src/backend/catalog/gp_toolkit.sql]: memory_redzone_limit would be better, let's keep it the same as syntax

Comment (pivotal-ning-yu) [inline src/backend/catalog/gp_toolkit.sql]: got it, will update soon.

**Comment by xiong-gang** [comment]
LGTM

**Comment by eurekaka** [comment]
LGTM

**Comment by pivotal-ning-yu** [comment]
Per our offline discussion I rebased the commits, now for each added/updated view there is a separate commit, totally three.

**Comment by pivotal-ning-yu** [comment]
Merged, thanks for your view.

File: src/backend/utils/adt/pgstatfuncs.c [hunk 9]
```diff
@@ -424,6 +424,13 @@ pg_stat_get_activity(PG_FUNCTION_ARGS)
 		if (nattr > 12)
 			TupleDescInitEntry(tupdesc, (AttrNumber) 13, "waiting_for", TEXTOID, -1, 0);
 
+		if (nattr > 13)
+		{
+			TupleDescInitEntry(tupdesc, (AttrNumber) 14, "rsgid", OIDOID, -1, 0);
+			TupleDescInitEntry(tupdesc, (AttrNumber) 15, "rsgname", TEXTOID, -1, 0);
+			TupleDescInitEntry(tupdesc, (AttrNumber) 16, "rsgqueueduration", INTERVALOID, -1, 0);
+		}
+
```
Comment (amilkh) [inline src/backend/utils/adt/pgstatfuncs.c]: Why are these protected by an if statement?

Comment (benchristel) [inline src/backend/utils/adt/pgstatfuncs.c]: Also, should these columns be renamed to `rsg_id`, `rsg_name`, `rsg_queued_duration` to match the other columns of `pg_stat_activity`?

Comment (pivotal-ning-yu) [inline src/backend/utils/adt/pgstatfuncs.c]: @amilkh thanks very much for the review. Just above here there were lines like below

```C
		if (nattr > 12)
			TupleDescInitEntry(tupdesc, (AttrNumber) 13, "waiting_for", TEXTOID, -1, 0);
```

I assume there is some historical reason for it to look like this. Maybe we can file another PR to cleanup these together. Please let me know your suggestions.

@benchristel thank you very much for the review, in fact I have the same concern with you, however in pg_stat_activity there are already two naming styles, one style is like `procpid`, `usesysid`, the other is like `sess_id`, `query_start`, so I didn't know which exact style should I follow. If the latter is better I could change it, please let me know your suggestions.

Comment (amilkh) [inline src/backend/utils/adt/pgstatfuncs.c]: About the if-blocking, @hlinnaka @asubramanya any thoughts?

About the naming style, we think that underscores are much better and easier to read (I confused rsgqueued with rsqqueued, the g vs q is hard to see)!
Also, `rg_id`, `rg_name`, `rg_queued_duration` may be better than using `rsg`?
Naming convention is inconsistent with the views also that are called `gp_resgroup_config` `gp_resgroup_status` in terms of `gp_resgroup` vs `rsg` vs `rg`

Comment (pivotal-ning-yu) [inline src/backend/utils/adt/pgstatfuncs.c]: @amilkh got your ideas, I'll improve the code following your suggestions. By the way, as the patches were already merged, I think a new PR is better for these changes, or do you think I should reopen this one?

Comment (pivotal-ning-yu) [inline src/backend/utils/adt/pgstatfuncs.c]: @amilkh I submitted the changes in a new PR: https://github.com/greenplum-db/gpdb/pull/1919 , could you please help to review it? Thanks in advance.

**Pull request closed without merging**